### ISSUE TITLE
Agree and document the best practices we recommend for how to write tests for nodes and pipelines

### ISSUE THREAD

**Issue by yetudada** [kedro-org/kedro#1271]
## Description

Our users have expressed a need to learn how to write tests for their nodes and pipelines. We encourage learning about software-engineering best practice and want to include a guide in our documentation for this. 

## Possible Implementation
This guide must focus on explaining:

- How `kedro test` works
- How to write a test for a node, using an example
- How to write a test for a pipeline, using an example

This guide can fall into our **Development** chapter. And remember to follow our guidelines for [contributing to the documentation](https://kedro.readthedocs.io/en/stable/14_contribution/04_documentation_contributor_guidelines.html).

**Comment by datajoely** [comment]
FYI @benhorsburgh

**Comment by deepyaman** [comment]
> How to write a test for a pipeline, using an example

I strongly believe that end-to-end pipeline tests should not be written as unit tests, as they retest the behavior of each contained node. Happy to discuss if contentious (I think a lot of people ascribe to running pipelines with data in pytest), and further willing to contribute to writing this part.

**Comment by datajoely** [comment]
@deepayman end-to-end pipeline tests can be written as integration tests right?

**Comment by datajoely** [comment]
If at all useful, please [steal what you can from here](https://protocol.quantumblack.com/docs/alloy/docs/source/03_tutorial/testing.html).

**Comment by daBlesr** [comment]
I'm not able to access those links, but in any case, in my experience writing tests for pipeline came to its full power when doing operations on Spark DataFrames. Usually, for pipelines, one test with data suffices. Here is a suggestion for what this page could look like using Spark DataFrames:

### Writing Tests for Nodes and Pipelines

In this section we introduce the way you write unit tests and integration tests for Nodes and Pipelines respectively. Each Node in a pipeline should have its own (parametrised) unit tests. Tests for pipelines on the other hand should test a number of sequential nodes. For such a test, the input of the first node is tested against the expected output of the last node. Imagine the following nodes, were we compute the total equity of a store's inventory:
```python
# pipelines/inventory/nodes.py
def process_total_price_per_product(inventory: DataFrame):
    return (
        inventory
            .withColumn("total_prices", col("in_store") * col("price"))
            .select("name", "total_price")
    )

def process_total_equity(total_price_per_product: DataFrame):
    return (
        total_price_per_product
            .groupBy()
            .sum('total_price')
            .first()
            [0]  # first col
    ) or 0.0
```
The pipeline:
```python
# pipelines/inventory/pipeline.py
def create_pipeline() -> Pipeline:
    return pipeline([
        node(process_total_price_per_product, inputs="inventory", outputs="total_price_per_product"),
        node(process_total_equity, inputs="total_price_per_product", outputs="total_equity")
    ])
```
## Unit Tests
We begin by writing the first unit test by spinning up two sample data sets from a CSV file:
*inventory*
```csv
name,price,in_store
apple,1.0,2
pear,1.5,5
banana,2.0,3
``` 
and *total_price_per_product*
```csv
name,total_price
apple,2.0
pear,7.5
banana,6.0
```
Since, we need to be able to read these CSV datasets into Spark DataFrames, we construct a Spark Session in the root of our testing package:
```python
# tests/conftest.py
@pytest.fixture
def spark(project_context) -> SparkSession:
    spark = (
        SparkSession.builder.master("local[*]")
            .appName("local-tests")
            .enableHiveSupport()
            .getOrCreate()
    )
    return spark
```
Now, we can create fixtures for loading these datasets:
```python
# tests/pipelines/inventory/conftest.py
@pytest.fixture
def inventory(spark):
    return spark.read.csv(test_data_inventory_path, header=True, inferSchema=True)

@pytest.fixture
def total_price_per_product(spark):
    return spark.read.csv(test_data_tpp_path, header=True, inferSchema=True)
```
In `test_nodes.py` we use the above fixtures to test our Nodes:
```python
# tests/pipelines/inventory/test_nodes.py
class TestInventoryNodes:

    def test_total_price_per_product(self, spark, inventory, total_price_per_product):
        processed_total_prices = process_total_price_per_product(inventory)

        assert 'total_price' in processed_total_prices.columns
        assert processed_total_prices.collect() == total_price_per_product.collect()

    def test_total_equity(self, spark, total_price_per_product):
        total_equity = process_total_equity(total_price_per_product)

        assert total_equity == 15.5
```
## Integration Test
As we have unit tests for each individual node in our Inventory pipeline, we would like to create an integration test for the pipeline as a whole. Again, we start by creating a number of fixtures in the root of the testing package:
```python
# tests/conftest.py
@pytest.fixture
def seq_runner():
    return SequentialRunner()

@pytest.fixture
def catalog():
    return DataCatalog()
```
The fixture guarantees we are working with a clean instance of the `DataCatalog` in our tests.
In the end we can write an integration test of the Inventory pipeline:
```python
# tests/pipelines/inventory/test_pipeline.py
class TestInventoryPipeline:

    def test_pipeline(
            self,
            spark,      # Spark Session
            seq_runner, # New SequentialRunner
            catalog,    # New Catalog
            inventory   # Inventory DataFrame
    ):

        # loading up the empty catalog for pipeline inputs.
        catalog.add("inventory", MemoryDataSet(inventory))

        # we want to do an integration test from start to this output
        pipeline = create_pipeline().to_outputs("total_equity")

        # run the pipeline
        output = seq_runner.run(pipeline, catalog)

        assert output['total_equity'] == 15.5
```
Make sure to add `to_outputs(..)` to `create_pipeline`, if you add nodes later to the pipeline, you'd still like to keep this integration test working. 

-----
Let me know what you guys think of this setup 😄

**Comment by datajoely** [comment]
This is excellent! Thank you @daBlesr

**Comment by analyticalmeanderings** [comment]
> We can use this [unit testing for data science](https://brix.quantumblack.com/post/tutorial-unit-testing-data-science-code-b2da2:latest) guide to help with this task.

Can you publish this publicly?

**Comment by datajoely** [comment]
This is actively being worked on :)

**Comment by stichbury** [comment]
We discussed this briefly in docs ticket grooming and best practices in testing is predicated on having a tech design session on how we recommend testing to users, since we cannot write about it until that is agreed.

**Comment by daBlesr** [comment]
A year, with lots of more experience on writing unit tests for Nodes, I have an update on this. For anyone interested, we noticed that our above approach was hard to maintain for three reasons:
* Using data from CSV files as an input to a unit tests causes the unit test to become _incomplete_ in their information. We noticed it is important to include all the necessary information to understand the data used during a unit test, and to understand what exactly is being asserted.
* By using CSV files it is hard to create multiple unit tests that tests different aspects of the code. Copying CSVs with slightly different data is not really a structured and maintainable approach.
* CSV files used for multiple tests as either input are output increases coupling, increasing maintenance overhead.

So how did improve our way of writing unit tests for Kedro nodes? Use factories! (note: our way of performing integration tests has been roughly unchanged). Examples below:

Our first small factory that creates a row in an inventory (similar to a row in the CSV file).
```
def product(name="apple", price=1.0, in_store=2):
    return Row(name=name, price=price, in_store=in_store)
```
Now we can create a fixture that returns a factory that can create an inventory on the go:
```
class TestInventoryNodes:

    @pytest.fixture
    def create_inventory(self, spark):
        def _create_inventory(inventory_rows: List[Row]):
            return spark.createDataFrame(inventory_rows, INVENTORY_SCHEMA)
        return _create_inventory
```

Here's the real deal, our test is **complete** in its information.
```
class TestInventoryNodes:
    ...

    def test_total_price_per_product(self, create_inventory):
        inventory = create_inventory([
            product(name="apple", price=1.0, in_store=2),
            product(name="pear", price=1.5, in_store=5),
            product(name="banana", price=2.0, in_store=3),
        ])
        processed_total_prices = process_total_price_per_product(inventory)

        assert processed_total_prices.collect() == [
            Row(name="apple", total_price=2.0),
            Row(name="pear", total_price=7.5),
            Row(name="banana", total_price=6.0),
        ]
```
By refraining from using CSV files, we can now much easier create more unit tests that test different aspects:
```
    def test_empty_inventory(self, create_inventory):
        inventory = create_inventory([])
        processed_total_prices = process_total_price_per_product(inventory)
        assert processed_total_prices.count() == 0
```
For the third reason mentioned above, using this approach we have easily decoupled unit tests, these unit tests are not at all related to each other anymore. 
```
class TestInventoryNodes:
    ...
    def test_total_equity(self, spark, create_product_equity_df):
        product_equity_df = create_product_equity_df([
            product_equity(name="apple", total_price=2.0),
            product_equity(name="pear", total_price=7.5),
        ])
        assert process_total_equity(product_equity_df) == 9.5
```
All in all, we have moved away from using CSV datasets for unit test input and outputs, and have been using factories since.

**Comment by astrojuanlu** [comment]
This came up again in conversations with internal users. As far as I understand, "tests for nodes and pipelines" requires at least 2 kinds of tests:

- Unit tests for node functions
- End to end tests for pipelines

We'd need to document an approach that is dataset-agnostic, but the examples should probably use dataframes.

**Comment by astrojuanlu** [comment]
This is consistently the most requested guidance from internal users. I'm raising its priority to High.

**Comment by datajoely** [comment]
We half finished an testing framework that I hoped to contribute before we had to deprioritise it. 

It's accessible internally here, but I would be very keen to see the core team absorb it.

- [Docs](https://brix.quantumblack.com/products/qblabs-monorepo/docs/docs/source/kedro-test/guide/guide.html)
- [Code](https://github.com/McK-Internal/qblabs-monorepo/tree/main/qblabs-monorepo/packages/incubation/kedro_test/src/kedro_test)

cc @benhorsburgh @anasthasiamanuQB

**Comment by astrojuanlu** [comment]
Shameless self-plug, tangentially related https://nbviewer.org/format/slides/github/poliastro/oscw2018-talk/blob/master/Talk.ipynb

Would love to see a focus on testing the node functions (unit tests), namely by (1) adding "validation tests" (given these inputs, these known outputs are returned) and (2) more sophisticated property-based testing, using hypothesis https://hypothesis.readthedocs.io/en/latest/quickstart.html, optionally adding pandera for schema testing and synthetic inputs generation https://pandera.readthedocs.io/en/stable/data_synthesis_strategies.html

**Comment by julnow** [comment]
> We half finished an testing framework that I hoped to contribute before we had to deprioritise it.
> 
> It's accessible internally here, but I would be very keen to see the core team absorb it.
> 
> * [Docs](https://brix.quantumblack.com/products/qblabs-monorepo/docs/docs/source/kedro-test/guide/guide.html)
> * [Code](https://github.com/McK-Internal/qblabs-monorepo/tree/main/qblabs-monorepo/packages/incubation/kedro_test/src/kedro_test)
> 
> cc @benhorsburgh @AnasthasiaManuQB

I'm really hoping you could make it public one day.

**Comment by datajoely** [comment]
Me too! Just needs a bit of love to bring it home

**Comment by AhdraMeraliQB** [comment]
Closed by #3782
